src/components/tl/mlx5/mcast/tl_mlx5_mcast_context.c

```diff
@@ -10,9 +10,241 @@
 #include <ucs/sys/string.h>
 #include "core/ucc_service_coll.h"
 #include "tl_mlx5.h"
+#include "tl_mlx5_mcast_helper.h"
+#include "tl_mlx5_mcast_rcache.h"
 
-ucc_status_t ucc_tl_mlx5_mcast_context_init(ucc_tl_mlx5_mcast_context_t    *context, /* NOLINT */
-                                            ucc_tl_mlx5_mcast_ctx_params_t *mcast_ctx_conf /* NOLINT */)
+#define UCC_TL_MLX5_MCAST_MAX_MTU_COUNT 5
+int mtu_lookup[UCC_TL_MLX5_MCAST_MAX_MTU_COUNT][2] =
+                        {{256, IBV_MTU_256}, {512, IBV_MTU_512},
+                        {1024, IBV_MTU_1024}, {2048, IBV_MTU_2048},
+                        {4096, IBV_MTU_4096}};
+
+ucc_status_t ucc_tl_mlx5_mcast_context_init(ucc_tl_mlx5_mcast_context_t    *context,
+                                            ucc_tl_mlx5_mcast_ctx_params_t *mcast_ctx_conf)
 {
+    ucc_status_t            status        = UCC_OK;
+    struct ibv_device     **device_list   = NULL;
+    struct ibv_device      *dev           = NULL;
+    char                   *devname       = NULL;
+    int                     is_ipv4       = 0;
+    struct sockaddr_in     *in_src_addr   = NULL;
+    struct rdma_cm_event   *revent        = NULL;
+    char                   *ib            = NULL;
+    char                   *ib_name       = NULL;
+    char                   *port          = NULL;
+    int                     active_mtu    = 4096;
+    int                     max_mtu       = 4096;
+    ucc_tl_mlx5_mcast_coll_context_t *ctx = NULL;
+
+    struct ibv_port_attr    port_attr;
+    struct ibv_device_attr  device_attr;
+    struct sockaddr_storage ip_oib_addr;
+    struct sockaddr_storage dst_addr;
+    int                     num_devices;
+    char                    addrstr[128];
+    ucc_tl_mlx5_context_t  *mlx5_ctx;
+    ucc_base_lib_t         *lib;
+    int                     i;
+    int                     user_provided_ib;
+    int                     ib_valid;
+    const char             *dst;
+
+    ctx = &(context->mcast_context);
+    memset(ctx, 0, sizeof(ucc_tl_mlx5_mcast_coll_context_t));
+    memcpy(&ctx->params, mcast_ctx_conf, sizeof(ucc_tl_mlx5_mcast_ctx_params_t));
+
+    mlx5_ctx = ucc_container_of(context, ucc_tl_mlx5_context_t, mcast);
+    lib      = mlx5_ctx->super.super.lib;
+    ctx->lib = lib;
+
+    device_list = ibv_get_device_list(&num_devices);
```

Comment:
i think this check should be common for TL MLX5, it's not multicast specific

Reply:
I think in the future when mlx5 TL is mature, we must unify all the contexts that we create in mlx5 TL. For now, we can continue with separate ctx. I will add this item in the to-do list for future.

Reply:
Sharing resources might be tricky since mlx5 alltoall need specific features e.g. shared pd. But at least checking ib devices should be easy